File: common/src/main/java/com/ishow/common/widget/StatusView.java
/*
 * Copyright (C) 2016 The yuhaiyang Android Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ishow.common.widget;

import android.animation.ObjectAnimator;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.TypedArray;
import android.graphics.drawable.AnimationDrawable;
import android.graphics.drawable.Drawable;
import androidx.annotation.DrawableRes;
import androidx.annotation.FloatRange;
import androidx.annotation.StringRes;
import androidx.core.content.ContextCompat;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.ishow.common.R;
import com.ishow.common.utils.log.LogUtils;
import com.ishow.common.widget.spinkit.SpinKitView;

/**
 * 一个状态显示的View
 */
@SuppressWarnings("unused")
public class StatusView extends FrameLayout implements View.OnClickListener {
    private static final String TAG = "StatusView";
    /**
     * 正在加载
     */
    public final static int STATUS_LOADING = 1 << 1;
    /**
     * 加载失败
     */
    public final static int STATUS_ERROR = STATUS_LOADING + 1;
    /**
     * 加载为空
     */
    public final static int STATUS_EMPTY = STATUS_LOADING + 2;

    public enum Which {
        Title, SubTitle, Reload
    }

    private ImageView mIconView;
    private SpinKitView mLoadingView;
    private TextView mTitleView;
    private TextView mSubTitleView;
    private TextView mReloadView;

    private ObjectAnimator mDismissAni;

    private int mLoadingDrawableId;
    private String mLoadingText;
    private ColorStateList mLoadingTextColor;
    private int mLoadingTextSize;

    private int mEmptyDrawableId;
    private String mEmptyText;
    private ColorStateList mEmptyTextColor;
    private int mEmptyTextSize;
    private String mEmptySubText;
    private ColorStateList mEmptySubTextColor;
    private int mEmptySubTextSize;

    private int mErrorDrawableId;
    private String mErrorText;
    private ColorStateList mErrorTextColor;
    private int mErrorTextSize;

    private String mErrorSubText;
    private ColorStateList mErrorSubTextColor;
    private int mErrorSubTextSize;

    private Drawable mReloadTextBackground;
    private String mReloadText;
    private ColorStateList mReloadTextColor;
    private int mReloadTextSize;

    private boolean isInterruptTouchEvent;
    private boolean isTitleClickable;
    private boolean isSubTitleClickable;
    private View mTopWeightView;
    private float mTopWeight;
    private View mBottomWeightView;
    private float mBottomWeight;
    private OnStatusViewListener mOnStatusViewListener;

    public StatusView(Context context) {
        this(context, null);
    }

    public StatusView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public StatusView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.StatusView, R.attr.statusStyle, R.style.Default_StatusView);
        mLoadingDrawableId = a.getResourceId(R.styleable.StatusView_loadingImage, -1);
        mLoadingText = a.getString(R.styleable.StatusView_loadingText);
        mLoadingTextColor = a.getColorStateList(R.styleable.StatusView_loadingTextColor);
        mLoadingTextSize = a.getDimensionPixelSize(R.styleable.StatusView_loadingTextSize, getDefaultTextSize());

        mEmptyDrawableId = a.getResourceId(R.styleable.StatusView_emptyImage, -1);
        mEmptyText = a.getString(R.styleable.StatusView_emptyText);
        mEmptyTextColor = a.getColorStateList(R.styleable.StatusView_emptyTextColor);
        mEmptyTextSize = a.getDimensionPixelSize(R.styleable.StatusView_emptyTextSize, getDefaultTextSize());
        mEmptySubText = a.getString(R.styleable.StatusView_emptySubText);
        mEmptySubTextColor = a.getColorStateList(R.styleable.StatusView_emptySubTextColor);
        mEmptySubTextSize = a.getDimensionPixelSize(R.styleable.StatusView_emptySubTextSize, getDefaultSubTextSize());

        mErrorDrawableId = a.getResourceId(R.styleable.StatusView_errorImage, -1);
        mErrorText = a.getString(R.styleable.StatusView_errorText);
        mErrorTextColor = a.getColorStateList(R.styleable.StatusView_errorTextColor);
        mErrorTextSize = a.getDimensionPixelSize(R.styleable.StatusView_errorTextSize, getDefaultTextSize());
        mErrorSubText = a.getString(R.styleable.StatusView_errorSubText);
        mErrorSubTextColor = a.getColorStateList(R.styleable.StatusView_errorSubTextColor);
        mErrorSubTextSize = a.getDimensionPixelSize(R.styleable.StatusView_errorSubTextSize, getDefaultSubTextSize());

        mReloadText = a.getString(R.styleable.StatusView_reloadText);
        mReloadTextColor = a.getColorStateList(R.styleable.StatusView_reloadTextColor);
        mReloadTextSize = a.getDimensionPixelSize(R.styleable.StatusView_reloadTextSize, getDefaultReloadTextSize());
        mReloadTextBackground = a.getDrawable(R.styleable.StatusView_reloadBackground);
        isInterruptTouchEvent = a.getBoolean(R.styleable.StatusView_interruptTouchEvent, true);
        isTitleClickable = a.getBoolean(R.styleable.StatusView_titleClickable, false);
        isSubTitleClickable = a.getBoolean(R.styleable.StatusView_subTitleClickable, false);
        mTopWeight = a.getFloat(R.styleable.StatusView_topWeight, 2.6F);
        mBottomWeight = a.getFloat(R.styleable.StatusView_topWeight, 5F);
        a.recycle();
        checkParams();
        initView();
    }


    private void initView() {
        LayoutInflater inflater = LayoutInflater.from(getContext());
        inflater.inflate(R.layout.widget_status_view, this, true);
        mIconView = findViewById(R.id.icon);
        mLoadingView = findViewById(R.id.status_loading);
        mTitleView = findViewById(R.id.title);
        if (isTitleClickable) mTitleView.setOnClickListener(this);

        mSubTitleView = findViewById(R.id.subTitle);
        if (isSubTitleClickable) mSubTitleView.setOnClickListener(this);
        mReloadView = findViewById(R.id.reload);
        mReloadView.setOnClickListener(this);


        mTopWeightView = findViewById(R.id.topWeight);
        updateWeight(mTopWeightView, mTopWeight);
        mBottomWeightView = findViewById(R.id.bottomWeight);
        updateWeight(mBottomWeightView, mBottomWeight);
    }


    private void checkParams() {
        final ColorStateList defaultTextColor = getDefaultTextColor();
        final ColorStateList defaultSubTextColor = getDefaultSubTextColor();

        if (mLoadingTextColor == null) mLoadingTextColor = defaultSubTextColor;
        if (mErrorTextColor == null) mErrorTextColor = defaultTextColor;
        if (mErrorSubTextColor == null) mErrorSubTextColor = defaultSubTextColor;
        if (mEmptyTextColor == null) mEmptyTextColor = defaultTextColor;
        if (mEmptySubTextColor == null) mEmptySubTextColor = defaultSubTextColor;
        if (mReloadTextColor == null) mReloadTextColor = getDefaultReloadTextColor();
    }


    public void showError() {
        showError(mReloadText, mErrorText, mErrorSubText, mErrorDrawableId);
    }

    public void showError(@StringRes int title, int icon) {
        Context context = getContext();
        String titleString = context.getString(title);
        showError(mReloadText, titleString, mErrorSubText, icon);
    }

    public void showError(@StringRes int title, @StringRes int subTitle, int icon) {
        Context context = getContext();
        String titleString = context.getString(title);
        String subTitleString = context.getString(subTitle);
        showError(mReloadText, titleString, subTitleString, icon);
    }

    public void showError(@StringRes int reload, @StringRes int title, @StringRes int subTitle, int icon) {
        Context context = getContext();
        String reloadString = context.getString(reload);
        String titleString = context.getString(title);
        String subTitleString = context.getString(subTitle);
        showError(reloadString, titleString, subTitleString, icon);
    }

    public void showError(String reload, String title, String subTitle, int icon) {
        setVisibility(VISIBLE);
        mIconView.setImageResource(icon);
        mIconView.setVisibility(VISIBLE);
        stopLoadingAnimation();
        mLoadingView.setVisibility(GONE);

        mTitleView.setTextColor(mErrorTextColor);
        mTitleView.setTextSize(TypedValue.COMPLEX_UNIT_PX, mErrorTextSize);
        setText(mTitleView, title);
        mSubTitleView.setTextColor(mErrorSubTextColor);
        mSubTitleView.setTextSize(TypedValue.COMPLEX_UNIT_PX, mErrorSubTextSize);
        setText(mSubTitleView, subTitle);
        mReloadView.setTextColor(mReloadTextColor);
        mReloadView.setTextSize(TypedValue.COMPLEX_UNIT_PX, mReloadTextSize);
        mReloadView.setBackground(mReloadTextBackground);
        setText(mReloadView, reload);

    }

    public void showLoading() {
        showLoading(mLoadingText, mLoadingDrawableId);
    }

    public void showLoading(@StringRes int text) {
        String subTitle = getContext().getString(text);
        showLoading(subTitle, mLoadingDrawableId);
    }

    public void showLoading(String title, @DrawableRes int drawable) {
        setVisibility(VISIBLE);
        mIconView.setVisibility(GONE);
        mLoadingView.setVisibility(VISIBLE);
        mTitleView.setTextColor(mLoadingTextColor);
        mTitleView.setTextSize(TypedValue.COMPLEX_UNIT_PX, mLoadingTextSize);
        setText(mTitleView, title);
        mSubTitleView.setVisibility(GONE);
        mReloadView.setVisibility(GONE);
    }


    public void showEmpty() {
        showEmpty(mEmptyText, mEmptySubText, mEmptyDrawableId);
    }

    public void showEmpty(@StringRes int text) {
        String title = getContext().getString(text);
        showEmpty(title, mEmptySubText, mEmptyDrawableId);
    }

    public void showEmpty(@StringRes int text, @DrawableRes int icon) {
        String title = getContext().getString(text);
        showEmpty(title, mEmptySubText, icon);
    }

    public void showEmpty(@StringRes int text, @StringRes int subText, @DrawableRes int icon) {
        String title = getContext().getString(text);
        String subTitle = getContext().getString(text);
        showEmpty(title, subTitle, icon);
    }

    public void showEmpty(String title, @DrawableRes int icon) {
        showEmpty(title, mEmptySubText, icon);
    }

    public void showEmpty(String title, String subTitle, @DrawableRes int icon) {
        setVisibility(VISIBLE);
        mIconView.setImageResource(icon);
        mIconView.setVisibility(VISIBLE);
        stopLoadingAnimation();
        mLoadingView.setVisibility(GONE);

        mTitleView.setTextColor(mEmptyTextColor);
        mTitleView.setTextSize(TypedValue.COMPLEX_UNIT_PX, mEmptyTextSize);
        setText(mTitleView, title);

        mSubTitleView.setTextColor(mEmptySubTextColor);
        mSubTitleView.setTextSize(TypedValue.COMPLEX_UNIT_PX, mEmptySubTextSize);
        setText(mSubTitleView, subTitle);

        mReloadView.setVisibility(GONE);
    }

    public void dismiss() {
        setVisibility(View.GONE);
    }

    public void cancelDismiss() {
        setVisibility(View.VISIBLE);
    }


    private void setText(TextView view, String text) {
        if (TextUtils.isEmpty(text)) {
            view.setVisibility(GONE);
        } else {
            view.setVisibility(VISIBLE);
        }
        view.setText(text);
    }

    @Override
    @SuppressLint("ClickableViewAccessibility")
    public boolean onTouchEvent(MotionEvent event) {
        if (isInterruptTouchEvent) {
            return true;
        } else {
            return super.onTouchEvent(event);
        }
    }

    @Override
    public void onClick(View v) {
        if (v == mTitleView) {
            notifyClickTitle();
        } else if (v == mSubTitleView) {
            notifyClickSubTitle();
        } else if (v == mReloadView) {
            notifyReload();
        }
    }

    private void startLoadingAnimation() {
        Drawable drawable = mIconView.getDrawable();
        if (!(drawable instanceof AnimationDrawable)) {
            LogUtils.e(TAG, "drawable is not AnimationDrawable");
            return;
        }

        ((AnimationDrawable) drawable).start();
    }

    private void stopLoadingAnimation() {
        mIconView.clearAnimation();
        Drawable drawable = mIconView.getDrawable();
        if (!(drawable instanceof AnimationDrawable)) {
            LogUtils.e(TAG, "drawable is not AnimationDrawable");
            return;
        }

        ((AnimationDrawable) drawable).stop();
    }

    public void setOnStatusViewListener(OnStatusViewListener callBack) {
        mOnStatusViewListener = callBack;
    }

    private void notifyClickTitle() {
        if (mOnStatusViewListener != null) {
            mOnStatusViewListener.onStatusClick(this, Which.Title);
        }
    }

    private void notifyClickSubTitle() {
        if (mOnStatusViewListener != null) {
            mOnStatusViewListener.onStatusClick(this, Which.SubTitle);
        }
    }

    private void notifyReload() {
        if (mOnStatusViewListener != null) {
            mOnStatusViewListener.onStatusClick(this, Which.Reload);
        }
    }

    public void setTitleClickable(boolean clickable) {
        mTitleView.setOnClickListener(clickable ? this : null);
    }

    public void setSubTitleClickable(boolean clickable) {
        mSubTitleView.setOnClickListener(clickable ? this : null);
    }

    private void updateWeight(View view, float weight) {
        ViewGroup.LayoutParams lp = view.getLayoutParams();
        if (lp instanceof LinearLayout.LayoutParams) {
            ((LinearLayout.LayoutParams) lp).weight = weight;
        }
    }

    public void setWeight(@FloatRange(from = 0F) float topWeight, @FloatRange(from = 0F)float bottomWeight) {
        mTopWeight = topWeight;
        updateWeight(mTopWeightView, mTopWeight);
        mBottomWeight = bottomWeight;
        updateWeight(mBottomWeightView, mBottomWeight);
    }

    public interface OnStatusViewListener {
        /**
         * 点击了副标题标题
         */
        void onStatusClick(View v, Which which);
    }

    /**
     * 默认字体大小
     */
    private int getDefaultTextSize() {
        return getContext().getResources().getDimensionPixelSize(R.dimen.H_title);
    }

    /**
     * 默认字体颜色
     */
    private ColorStateList getDefaultTextColor() {
        return ContextCompat.getColorStateList(getContext(), R.color.text_grey_normal);
    }

    /**
     * 默认副标题字体大小
     */
    private int getDefaultSubTextSize() {
        return getContext().getResources().getDimensionPixelSize(R.dimen.J_title);
    }

    /**
     * 默认副标题字体颜色
     */
    private ColorStateList getDefaultSubTextColor() {
        return ContextCompat.getColorStateList(getContext(), R.color.text_grey_light_more_normal);
    }

    /**
     * 默认重新加载字体大小
     */
    private int getDefaultReloadTextSize() {
        return getContext().getResources().getDimensionPixelSize(R.dimen.H_title);
    }

    /**
     * 默认重新加载字体颜色
     */
    private ColorStateList getDefaultReloadTextColor() {
        return ContextCompat.getColorStateList(getContext(), R.color.text_grey_normal);
    }
}

File: common/src/main/java/com/ishow/common/widget/StatusView.kt
/*
 * Copyright (C) 2016 The yuhaiyang Android Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ishow.common.widget

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.text.TextUtils
import android.util.AttributeSet
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.TextView
import androidx.annotation.FloatRange
import androidx.core.content.ContextCompat
import com.ishow.common.R
import com.ishow.common.utils.log.LogUtils
import kotlinx.android.synthetic.main.widget_status_view.view.*

/**
 * 一个状态显示的View
 */
class StatusView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) :
    FrameLayout(context, attrs, defStyleAttr), View.OnClickListener {

    var loadingText: String?
    var loadingTextColor: ColorStateList? = null
    var loadingTextSize: Int

    @Suppress("MemberVisibilityCanBePrivate")
    var emptyDrawableId: Int
    var emptyText: String?
    var emptyTextColor: ColorStateList? = null
    var emptyTextSize: Int
    var emptySubText: String?
    var emptySubTextColor: ColorStateList? = null
    var emptySubTextSize: Int

    @Suppress("MemberVisibilityCanBePrivate")
    var errorDrawableId: Int
    var errorText: String?
    var errorTextColor: ColorStateList? = null
    var errorTextSize: Int

    var errorSubText: String?
    var errorSubTextColor: ColorStateList? = null
    var errorSubTextSize: Int

    @Suppress("MemberVisibilityCanBePrivate")
    var reloadTextBackground: Drawable?
    var reloadText: String?
    var reloadTextColor: ColorStateList? = null
    var reloadTextSize: Int

    private val isInterruptTouchEvent: Boolean
    private val isTitleClickable: Boolean
    private val isSubTitleClickable: Boolean

    private var topWeight: Float = 0.toFloat()
    private var bottomWeight: Float = 0.toFloat()

    private var statusViewListener: OnStatusViewListener? = null
    private var statusViewBlock: ((v: View, which: Which) -> Unit)? = null

    private var loadingTagList: MutableList<String>? = null
    private var hasError: Boolean = false
    /**
     * 默认字体大小
     */
    private val defaultTextSize: Int
        get() = context.resources.getDimensionPixelSize(R.dimen.H_title)

    /**
     * 默认字体颜色
     */
    private val defaultTextColor: ColorStateList?
        get() = ContextCompat.getColorStateList(context, R.color.text_grey_light)

    /**
     * 默认副标题字体大小
     */
    private val defaultSubTextSize: Int
        get() = context.resources.getDimensionPixelSize(R.dimen.J_title)

    /**
     * 默认副标题字体颜色
     */
    private val defaultSubTextColor: ColorStateList?
        get() = ContextCompat.getColorStateList(context, R.color.text_grey_light_more)

    /**
     * 默认重新加载字体大小
     */
    private val defaultReloadTextSize: Int
        get() = context.resources.getDimensionPixelSize(R.dimen.H_title)

    /**
     * 默认重新加载字体颜色
     */
    private val defaultReloadTextColor: ColorStateList?
        get() = ContextCompat.getColorStateList(context, R.color.text_grey)

    enum class Which {
        Title, SubTitle, Reload
    }

    init {
        val a = context.obtainStyledAttributes(
            attrs,
            R.styleable.StatusView,
            R.attr.statusStyle,
            R.style.Default_StatusView
        )

        topWeight = a.getFloat(R.styleable.StatusView_topWeight, 2.6F)
        bottomWeight = a.getFloat(R.styleable.StatusView_bottomWeight, 5F)

        loadingText = a.getString(R.styleable.StatusView_loadingText)
        loadingTextColor = a.getColorStateList(R.styleable.StatusView_loadingTextColor)
        loadingTextSize = a.getDimensionPixelSize(R.styleable.StatusView_loadingTextSize, defaultTextSize)

        emptyDrawableId = a.getResourceId(R.styleable.StatusView_emptyImage, -1)
        emptyText = a.getString(R.styleable.StatusView_emptyText)
        emptyTextColor = a.getColorStateList(R.styleable.StatusView_emptyTextColor)
        emptyTextSize = a.getDimensionPixelSize(R.styleable.StatusView_emptyTextSize, defaultTextSize)
        emptySubText = a.getString(R.styleable.StatusView_emptySubText)
        emptySubTextColor = a.getColorStateList(R.styleable.StatusView_emptySubTextColor)
        emptySubTextSize = a.getDimensionPixelSize(R.styleable.StatusView_emptySubTextSize, defaultSubTextSize)

        errorDrawableId = a.getResourceId(R.styleable.StatusView_errorImage, -1)
        errorText = a.getString(R.styleable.StatusView_errorText)
        errorTextColor = a.getColorStateList(R.styleable.StatusView_errorTextColor)
        errorTextSize = a.getDimensionPixelSize(R.styleable.StatusView_errorTextSize, defaultTextSize)
        errorSubText = a.getString(R.styleable.StatusView_errorSubText)
        errorSubTextColor = a.getColorStateList(R.styleable.StatusView_errorSubTextColor)
        errorSubTextSize = a.getDimensionPixelSize(R.styleable.StatusView_errorSubTextSize, defaultSubTextSize)

        reloadText = a.getString(R.styleable.StatusView_reloadText)
        reloadTextColor = a.getColorStateList(R.styleable.StatusView_reloadTextColor)
        reloadTextSize = a.getDimensionPixelSize(R.styleable.StatusView_reloadTextSize, defaultReloadTextSize)
        reloadTextBackground = a.getDrawable(R.styleable.StatusView_reloadBackground)

        isInterruptTouchEvent = a.getBoolean(R.styleable.StatusView_interruptTouchEvent, true)
        isTitleClickable = a.getBoolean(R.styleable.StatusView_titleClickable, false)
        isSubTitleClickable = a.getBoolean(R.styleable.StatusView_subTitleClickable, false)

        a.recycle()
        checkParams()
        initView()
    }


    private fun initView() {
        val inflater = LayoutInflater.from(context)
        inflater.inflate(R.layout.widget_status_view, this, true)
        setTitleClickable(isTitleClickable)
        setSubTitleClickable(isSubTitleClickable)
        reloadButton.setOnClickListener(this)
        setWeight(topWeight, bottomWeight)
    }

    private fun checkParams() {
        if (loadingTextColor == null) loadingTextColor = defaultSubTextColor

        if (errorTextColor == null) errorTextColor = defaultTextColor
        if (errorSubTextColor == null) errorSubTextColor = defaultSubTextColor

        if (emptyTextColor == null) emptyTextColor = defaultTextColor
        if (emptySubTextColor == null) emptySubTextColor = defaultSubTextColor

        if (reloadTextColor == null) reloadTextColor = defaultReloadTextColor
    }


    fun showError() {
        hasError = true
        if (loadingTagList != null && loadingTagList!!.isNotEmpty()) {
            return
        }

        visibility = View.VISIBLE
        imageView.setImageResource(errorDrawableId)
        imageView.visibility = View.VISIBLE
        loadingView.visibility = View.GONE
        loadingView.visibility = View.GONE

        titleView.setTextColor(errorTextColor)
        titleView.setTextSize(TypedValue.COMPLEX_UNIT_PX, errorTextSize.toFloat())
        setText(titleView, errorText)
        subTitleView.setTextColor(errorSubTextColor)
        subTitleView.setTextSize(TypedValue.COMPLEX_UNIT_PX, errorSubTextSize.toFloat())
        setText(subTitleView, errorSubText)
        reloadButton.setTextColor(reloadTextColor)
        reloadButton.setTextSize(TypedValue.COMPLEX_UNIT_PX, reloadTextSize.toFloat())
        reloadButton.background = reloadTextBackground
        setText(reloadButton, reloadText)
    }

    fun showLoading(loadingTag: String? = null) {
        loadingTag?.let {
            if (loadingTagList == null) {
                loadingTagList = mutableListOf()
            }
            loadingTagList?.add(loadingTag)
        }

        visibility = View.VISIBLE
        imageView.visibility = View.GONE
        loadingView.visibility = View.VISIBLE

        titleView.setTextColor(loadingTextColor)
        titleView.setTextSize(TypedValue.COMPLEX_UNIT_PX, loadingTextSize.toFloat())
        setText(titleView, loadingText)

        subTitleView.visibility = View.GONE
        reloadButton.visibility = View.GONE
    }

    fun showEmpty() {
        visibility = View.VISIBLE
        imageView.setImageResource(emptyDrawableId)
        imageView.visibility = View.VISIBLE
        loadingView.visibility = View.GONE

        titleView.setTextColor(emptyTextColor)
        titleView.setTextSize(TypedValue.COMPLEX_UNIT_PX, emptyTextSize.toFloat())
        setText(titleView, emptyText)

        subTitleView.setTextColor(emptySubTextColor)
        subTitleView.setTextSize(TypedValue.COMPLEX_UNIT_PX, emptySubTextSize.toFloat())
        setText(subTitleView, emptySubText)

        reloadButton.visibility = View.GONE
    }

    fun dismiss(loadingTag: String? = null) {
        loadingTagList?.remove(loadingTag)
        if (loadingTag == null || loadingTagList.isNullOrEmpty()) {
            if (hasError) {
                showError()
            } else {
                visibility = View.GONE
            }
        }
    }

    private fun setText(view: TextView?, text: String?) {
        if (TextUtils.isEmpty(text)) {
            view?.visibility = View.GONE
        } else {
            view?.visibility = View.VISIBLE
        }
        view?.text = text
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        return if (isInterruptTouchEvent) {
            true
        } else {
            super.onTouchEvent(event)
        }
    }

    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.titleView -> notifyClickTitle()
            R.id.subTitleView -> notifyClickSubTitle()
            R.id.reloadButton -> notifyReload()
        }
    }

    fun setOnStatusViewListener(callBack: OnStatusViewListener) {
        statusViewListener = callBack
    }

    @Suppress("unused")
    fun setOnStatusViewBlock(block: ((v: View, which: Which) -> Unit)?) {
        statusViewBlock = block
    }

    private fun notifyClickTitle() {
        statusViewListener?.onStatusClick(this, Which.Title)
        statusViewBlock?.let { it(this, Which.Title) }
    }

    private fun notifyClickSubTitle() {
        statusViewListener?.onStatusClick(this, Which.SubTitle)
        statusViewBlock?.let { it(this, Which.SubTitle) }
    }

    private fun notifyReload() {
        statusViewListener?.onStatusClick(this, Which.Reload)
        statusViewBlock?.let { it(this, Which.Reload) }
    }

    fun setTitleClickable(clickable: Boolean) {
        titleView.setOnClickListener(if (clickable) this else null)
    }

    fun setSubTitleClickable(clickable: Boolean) {
        subTitleView.setOnClickListener(if (clickable) this else null)
    }

    private fun updateWeight(view: View?, weight: Float) {
        view?.let {
            val lp = it.layoutParams
            if (lp is LinearLayout.LayoutParams) {
                lp.weight = weight
            }
        }
    }

    fun setWeight(@FloatRange(from = 0.0) topWeight: Float, @FloatRange(from = 0.0) bottomWeight: Float) {
        this.topWeight = topWeight
        updateWeight(topWeightView, this.topWeight)
        this.bottomWeight = bottomWeight
        updateWeight(bottomWeightView, this.bottomWeight)
    }


    interface OnStatusViewListener {
        /**
         * 点击了副标题标题
         */
        fun onStatusClick(v: View, which: Which)
    }

    companion object {
        private const val TAG = "StatusView"
        /**
         * 正在加载
         */
        const val STATUS_LOADING = 1 shl 1
        /**
         * 加载失败
         */
        const val STATUS_ERROR = STATUS_LOADING + 1
        /**
         * 加载为空
         */
        const val STATUS_EMPTY = STATUS_LOADING + 2


        /**
         * 添加到attachView
         */
        fun attachToActivity(activity: Activity?, topMargin: Int, color: Int = Color.WHITE): StatusView? {
            if (activity == null || activity.isFinishing) {
                return null
            }
            val view: StatusView? = activity.findViewById(R.id.statusView)
            if (view != null) {
                LogUtils.i(TAG, "already add")
                return view
            }

            val statusView = StatusView(activity)
            statusView.setBackgroundColor(color)
            statusView.id = R.id.statusView
            val root = activity.findViewById<ViewGroup>(android.R.id.content)

            val layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
            layoutParams.topMargin = topMargin
            root.addView(statusView, layoutParams)
            return statusView
        }


        /**
         * 添加到attachView
         */
        fun attach(view: ViewGroup, topMargin: Int, color: Int = Color.WHITE): StatusView? {
            val status: StatusView? = view.findViewById(R.id.statusView)
            if (status != null) {
                LogUtils.i(TAG, "already add")
                return status
            }

            val statusView = StatusView(view.context)
            statusView.setBackgroundColor(color)
            statusView.id = R.id.statusView

            val layoutParams =
                MarginLayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
            layoutParams.topMargin = topMargin
            view.addView(statusView, layoutParams)
            return statusView
        }

        fun detach(view: ViewGroup) {
            val status: StatusView? = view.findViewById(R.id.statusView)
            status?.let { view.removeView(it) }
        }
    }
}
